Automattic/wp-calypso
# chore: Migrate to Yarn v3

## Yarn v3

![image](https://user-images.githubusercontent.com/975703/128009143-df97cd65-12df-4cfc-bf9c-4ef1a024dc49.png)

### How to test

Check out this branch, delete node_modules (including nested ones, `rm -fr **/node_modules`), run your usual `yarn` commands and check everything works.

### Notable changes

#### Skipping yarn v2

Uses yarn v3 instead of yarn v2. Yarn v2 has some bugs (https://github.com/yarnpkg/berry/issues/2603, https://github.com/yarnpkg/berry/issues/2204) that make hiding logs quite difficult. Using `3.0.0` alleviates the problem.

#### Peer dependencies (direct)

Yarn made explicit that peer dependencies are not transitive. If `Package A` declares a `peerDependency` on `lib`, then consumers of `Package A` must either provide `lib` or declare it as a `peerDependency` (https://dev.to/arcanis/implicit-transitive-peer-dependencies-ed0).

I've added them to `package.json` as needed. The rule I followed is expose them as `peerDependencies` when the package is meant to be used by somebody else (e.g. `packages/*`), and provide them as `dependencies` when the package is "final" (e.g. `client/package.json`, `apps/*`)

#### Peer dependencies (transitive)

As above, but for transitive dependencies. In that case we have to "override" the package definition of the offending package to re-declare their peer dependencies. This is done in `.yarnrc.yml`. See https://yarnpkg.com/getting-started/migration#fix-dependencies-with-packageextensions for more info.

You'll notice there are lots of `@wordpress/*` packages. I believe that most of them have changed their dependencies in the latest versions (eg: no more `@wp-g2/create-styles`). The good news is that Yarn will print a message if an override dependency doesn't apply anymore, so those should be somewhat easy to clean.

#### nohoist

We were using Yarn v1 `nohoist` to avoid some problems with dependencies:
* `testarmada-magellan` is expecting `../mocha` to exist. We used to have two copies of mocha, `node_modules/mocha` and `test/e2e/node_modules/mocha`. If `testarmada-magellan` was hoisted to the root, it would pick the wrong version of mocha. This has been corrected for some time (we don't have two copies of `mocha` anymore) so `testarmada-magellan` can be hoisted.
* Same story for `@wordpress/edit-post` and `@wordpress/interface` (see https://github.com/Automattic/wp-calypso/pull/46515#discussion_r509082889). Since then, we only have one version of `@wordpress/interface`, so this is safe to hoist.

As such, we no longer need to tweak the hoisting algorithm.

#### swiper patch

`swiper@4.5.1` can't be installed using Yarn because their `postinstall` script (it has an unescaped quote). It has been raised in https://github.com/yarnpkg/berry/issues/1356, but no solution yet. I used the [patch protocol](https://yarnpkg.com/features/protocols#patch) to remove the `postinstall` script.

#### magellan patch

`testarmada-magellan@^11.0.10` tries to load a package by manually composing the path to `node_moudules` and it fails in a monorepo context (https://github.com/TestArmada/magellan/blob/master/src/cli.js#L85). The project is abandoned there is no way to send a patch upstream. Given we want to deprecate Magellan and move to Jest I figured the easiest way is to patch it. An alternative solution would be to fork it.

#### Calypso doctor

Previously we were running `@automattic/calypso-doctor` on `postinstall`, which runs every time a developer runs `yarn`. However, with yarn v3 this is not the case anymore, as it won't run if there are not changes in `package.json`. Also it doesn't display the output to the developer. We need to find a different way to run it periodically, likely a yarn plugin. I think we can disable it for now and work on the plugin on a separate PR.

#### `package.json` cleanup

Looks like `yarn` lints each `package.json` automatically. This has caused a few minor changes in those files (removing empty sections, sort dependencies alphabetically...)

#### Deprecated pre/post scripts

Yarn v3 won't automatically run `preX` or `postX` when running `yarn run X`. As such, I've migrated a few of those scripts to call the pre/post variants explicitly.

#### Dependencies for scripts

If a package in the repo has a script that calls `X`, then it must have a dependency on `X`. Yarn v1 was more lenient: as long as `X` was found in a parent's `node_modules` it was ok, but this is not the case in v3. Specifically, every package that runs `tsc` needs to have a dependency on `typescript`.

#### Duplicated packages

Yarn v3 has a plugin to detect duplicated packages natively: `yarn dedupe`. No need to use `npx yarn-deduplicate` anymore. I've adapted build scripts to use the new plugin.

#### Build caches

I had problems when running this branch if we reuse a cached `node_modules` from previous builds, likely because that `node_modules` was produced with yarn v1. I've disabled the caches for now, to be enabled once this lands in `trunk`.

#### Yarn shell

Yarn v3 run all scripts in a [Normalized Shell](https://dev.to/arcanis/introducing-yarn-2-4eh1#normalized-shell). I haven't found documentation about it, but after a bit trial and error, looks like we need to quote everything that has an `*` (eg:  commands like `npx rimraf dist/*`). I haven't done an exhaustive search to fix all scripts in our repo, I focused on the ones used by CI.

#### Newspack-blocks

`newspack-blocks#v1.30.0` is problematic because it depends on a version of `node-sass` (via transitive dependency of `@automattic/calypso-build`) that can be built only with Node 14. I've updated it to `v1.33.0`, a version that includes an update `calypso-build` (https://github.com/Automattic/newspack-blocks/pull/820).

#### Replaced lerna with `workspaces foreach`

In yarn v1 we used `lerna` to run scripts on all packages in the monorepo, specifically `lerna run prepare --stream` and `lerna run clean --stream`. This will fail in yarn v3 because `lerna` is not happy with the patch protocol (see https://github.com/lerna/lerna/issues/2610). I've replaced those scripts with [`yarn workspaces foreach`](https://yarnpkg.com/cli/workspaces/foreach) which does basically the same.

The downside is `workspaces foreach` is a bit slower: `clean` takes ~33s with `workspaces foreach`, ~20s with `lerna`. `prepare` is about the same in both (~5s). I think the main reason is that `lerna` spins a process for each package at once, while `workspaces foreach` has concurrency limits.

#### Desktop

`./desktop` has it own `yarn.lock` and for all intent and purposes, it is a separate dependency tree isolated from the root package. However, it was sharing yarn config in `./yarnrc`. In order to keep this PR small, I've configured `./desktop` to have its own `./desktop/.yarnrc` with Yarn v1. We can migrate desktop to yarn v3 in a separate PR.


### Pending

* [x] Deal with deprecated pre/post scripts (https://yarnpkg.com/advanced/lifecycle-scripts)
* [x] Use `packages/effective-module-tree` to verify the dependency tree has not changed unexpectedly.
* [x] Fix TS projects. Right now they fail to compile.
* [x] Deal with deprecated `nohoist`
* [x] Fix duplicate package detection
* [x] Release a new version of newspack-blocks without `node-sass`

### Post migration tasks
* [x] Fix calypso-build dependencies (#55181)
* [ ] Migrate from `npx` to `yarn dlx` (#55182)
* [ ] Standardize on `yarn run <script>`
* [ ] Figure out when/how to run Calypso Doctor (check with @sirreal, I think they wrote a plugin for Redpop)
* [x] Re-enable node_modules cache (or potentially rework this system to only save yarn cache, see @merceyz comments)
* [x] Drop lerna. AFAIK we were only using it to run scripts in all packages, and yarn v3 has `yarn workspaces foreach` for that purpose.
* [ ] Migrate desktop to the main dependency tree (#55184)
